src/main/java/org/apache/mesos/hdfs/Scheduler.java

```diff
@@ -191,197 +371,190 @@ private ExecutorInfo createExecutor(String taskIdName, String nodeName, String e
-          driver.declineOffer(offerID);
-        }
-        pendingOffers.clear();
+    boolean launch = false;
+    List<String> deadNameNodes = persistentState.getDeadNameNodes();
+
+    if (deadNameNodes.isEmpty()) {
+      if (liveState.getNameNodeSize() >= HDFSConstants.TOTAL_NAME_NODES) {
+        log.info(String.format("Already running %s namenodes", HDFSConstants.TOTAL_NAME_NODES));
+      } else if (persistentState.nameNodeRunningOnSlave(offer.getHostname())) {
+        log.info(String.format("Already running namenode on %s", offer.getHostname()));
+      } else if (persistentState.dataNodeRunningOnSlave(offer.getHostname())) {
+        log.info(String.format("Cannot colocate namenode and datanode on %s", offer.getHostname()));
+      } else if (!persistentState.journalNodeRunningOnSlave(offer.getHostname())) {
+        log.info(String.format("We need to coloate the namenode with a journalnode and there is"
+            + "no journalnode running on this host. %s", offer.getHostname()));
+      } else {
+        launch = true;
       }
-      return;
+    } else if (deadNameNodes.contains(offer.getHostname())) {
+      launch = true;
     }
+    if (launch) {
+      launchNode(
+          driver,
+          offer,
+          HDFSConstants.NAME_NODE_ID,
+          Arrays.asList(HDFSConstants.NAME_NODE_ID, HDFSConstants.ZKFC_NODE_ID),
+          HDFSConstants.NAME_NODE_EXECUTOR_ID);
+      return true;
+    }
+    return false;
+  }
 
-    List<Offer> remainingOffers = new ArrayList<>();
-    remainingOffers.addAll(offers);
-
-    if (initializingCluster) {
-      log.info(String.format("Declining remaining %d offers pending initialization",
-          remainingOffers.size()));
-      for (Offer offer : remainingOffers) {
-        driver.declineOffer(offer.getId());
-      }
-      return;
+  private boolean tryToLaunchDataNode(SchedulerDriver driver, Offer offer) {
+    if (offerNotEnoughResources(offer, conf.getDataNodeCpus(), conf.getDataNodeHeapSize())) {
+      log.info("Offer does not have enough resources");
+      return false;
     }
 
-    offers = remainingOffers;
-    remainingOffers = new ArrayList<>();
+    boolean launch = false;
+    List<String> deadDataNodes = persistentState.getDeadDataNodes();
 
-    // Check to see if we can launch some DataNodes
-    for (Offer offer : offers) {
-      if (liveState.notInDfsHosts(offer.getSlaveId().getValue())) {
-        launchDataNode(driver, offer);
+    if (deadDataNodes.isEmpty()) {
+      if (persistentState.dataNodeRunningOnSlave(offer.getHostname())
+          || persistentState.nameNodeRunningOnSlave(offer.getHostname())
+          || persistentState.journalNodeRunningOnSlave(offer.getHostname())) {
+        log.info(String.format("Already running hdfs task on %s", offer.getHostname()));
       } else {
-        remainingOffers.add(offer);
+        launch = true;
       }
+    } else if (deadDataNodes.contains(offer.getHostname())) {
```

Comment:
Again, if that DN host never comes up, we'll be stuck waiting for it and unable to launch any more DNs. We should prefer to relaunch a DN on this host if possible, but after some timeout move on to other nodes.
But if we're trying to launch a DN on every host anyways, why do you even need to check for deadDNs, or relaunch those first?